planner/core/find_best_task.go

```diff
@@ -624,14 +624,29 @@ func (ds *DataSource) skylinePruning(prop *property.PhysicalProperty) []*candida
 	}
 
 	if ds.ctx.GetSessionVars().GetAllowPreferRangeScan() && len(candidates) > 1 {
-		// remove the table/index full scan path
-		for i, c := range candidates {
-			for _, ran := range c.path.Ranges {
-				if ran.IsFullRange() {
-					candidates = append(candidates[:i], candidates[i+1:]...)
-					return candidates
+		// For those paths which are neither TiFlash-path nor forced-path, we remove table/index full scan paths if there
+		// exists some range scan path.
+		// For TiFlash-path or forced-path, we just keep them.
+		preferredPaths := make([]*candidatePath, 0, len(candidates))
+		var hasRangeScanPath bool
+		for _, c := range candidates {
+			if c.path.Forced || c.path.StoreType == kv.TiFlash {
+				preferredPaths = append(preferredPaths, c)
```

Comment:
Seems we should also set `hasRangeScanPath` to `true` here?

Reply:
If we have `Forced` paths, which means that the user intends to use certain paths, IMHO, in the situation `tidb_opt_prefer_range_scan` should not interfere too much with the user's choice. `TiFlash` paths are different from `TiKV` paths  so we ignore them when preferring range scan.